### PR TITLE
Add script to download canvas submission data

### LINKED ISSUE #15: Download canvas submission data

### PR DESCRIPTION
**Pull request by MtheDV**
Closes #15

### CODE REVIEW

**Review by carson-ricca — APPROVED** [review]
Looks good to me, if it's been tested and does what is expected with the submissions then it should be good to merge.

**Review by carson-ricca — DISMISSED** [review]

File: scripts/canvas_submission_retrieval/canvas_submission_retrieval.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Download Course Submission Data v1
```
Comment (Opeyem1a) [inline scripts/canvas_submission_retrieval/canvas_submission_retrieval.md:1]: I feel like our .md files are so in depth and they don't need to be - take a look at #18 and see the updated .md files for each script. There's no need to document file structure since again, we should just explain the goals of the script in plain english

File: scripts/canvas_submission_retrieval/canvas_submission_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        else:
+            return super().default(z)
+
+
+def canvas_submission_retrieval(course_id, access_token):
+    canvas = Canvas('https://canvas.ubc.ca', access_token)
+    assignments = canvas.get_course(course_id).get_assignments()
+    for assignment in assignments:
+        submissions = assignment.get_submissions()
+        with open(f"../../data/{course_id}_{assignment.id}.json", "w") as f:
```
Comment (Opeyem1a) [inline scripts/canvas_submission_retrieval/canvas_submission_retrieval.py:21]: this directory path (`../../data/`) should be a parameter of this function. We run these files from entrypoint files located elsewhere in the project, so relative pathing doesn't work consistently like this

File: scripts/canvas_submission_retrieval/canvas_submission_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            json_submissions = []
+            for submission in submissions:
+                submission_dict = submission.__dict__
+                submission_dict.pop('_requester')
+                json_submissions.append(json.dumps(submission_dict, indent=4, cls=DateTimeEncoder))
+            f.write('[' + ','.join(json_submissions) + ']')
+
+
+if __name__ == '__main__':
```
Comment (Opeyem1a) [inline scripts/canvas_submission_retrieval/canvas_submission_retrieval.py:30]: I think it'd be best to make an entrypoint file in the root directory that calls this method instead of using the `if __name__ == '__main__':` pattern here

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: submission_retrieval.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+EXPORT_DIR = os.path.join('data', 'canvas_submission')
+CANVAS_ACCESS_TOKEN = sys.argv[1]
+CANVAS_COURSE_ID = sys.argv[2]
+GRADE_BOOK_CSV_PATH = 'raw_data/2021-09-13T1128_Grades-COSC_341_COSC_541_101_2020W.csv'
+
+if __name__ == '__main__':
+
+    key_path = generate_key(GRADE_BOOK_CSV_PATH, KEY_DIR, KEY_EXPORT_FILENAME)
+    encoder = Encoder(key_path)
```
Comment (Opeyem1a) [inline submission_retrieval.py:19]: we shouldn't generate a key in this script, so for now make a constant that hardcodes the path to `keys/Key,csv` and just pass that directly into the encoder

File: submission_retrieval.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+from scripts import canvas_submission_retrieval, generate_key
+import sys
+
+from util import Encoder
+
+KEY_DIR = 'keys'
+KEY_EXPORT_FILENAME = 'Key'
+
+EXPORT_DIR = os.path.join('data', 'canvas_submission')
```
Comment (Opeyem1a) [inline submission_retrieval.py:11]: if you're just setting this to a constant, doing this is equivalent to hardcoding

File: submission_retrieval.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+CANVAS_ACCESS_TOKEN = sys.argv[1]
+CANVAS_COURSE_ID = sys.argv[2]
```
Comment (Opeyem1a) [inline submission_retrieval.py:13]: Add a section in the README to explain exactly how to call this script (we'll figure out better solutions later)
Something like "Run `python canvas_submission_retrieval <ACCESS_TOKEN> <COURSE_ID>`"

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: submission_retrieval.py (hunk 5 above)
Comment (Rick-Feng-u) [reply submission_retrieval.py:11]: EXPORT_DIR is supposed to be a hardcoded path. We decided to use os.path.join is because we could easily add more directories in if it is required.

**Review by Opeyem1a — DISMISSED** [review]

**Review by aadiraju — DISMISSED** [review]
Looks fine. I think we need to think about how we can avoid hardcoding paths for our files into each script, and maybe export the variables so there is only one global pathset. Specifically, I mean the `KEY_PATH`

**Comment by Rick-Feng-u** [comment]
> Looks fine. I think we need to think about how we can avoid hardcoding paths for our files into each script, and maybe export the variables so there is only one global pathset. Specifically, I mean the `KEY_PATH`

Maybe we should consider env files?

**Comment by Opeyem1a** [comment]
We'll be adding .env files in a PR coming in soon 👍

**Review by Opeyem1a — APPROVED** [review]
